### PR TITLE
Shooter

### PR DESCRIPTION
**Pull request by minhnguyenbhs**

### CODE REVIEW

**Comment by minhnguyenbhs** [comment]
I'm not sure that further work on this would make the sim actually more meaningful without testing on the real robot, so could you review this and see what I should edit? Thank you!

File: src/main/java/frc/robot/subsystems/scoring/ScoringSubsystem.java [hunk 1]
```diff
@@ -0,0 +1,732 @@
...
+                                        - ScoringConstants.shooterLowerVelocityMarginRPM);
+        boolean aimReady =
+                Math.abs(aimerInputs.aimAngleRad - aimerInputs.aimGoalAngleRad)
+                                < aimerAngleTolerance.getValue(distanceToGoal)
+                        && Math.abs(aimerInputs.aimVelocityErrorRadPerSec)
+                                < ScoringConstants.aimAngleVelocityMargin;
+        boolean driveReady = driveAllignedSupplier.get();
+        boolean fieldLocationReady = true;
+
+        //TODO FIX
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/scoring/ScoringSubsystem.java:250]: Should this be fixed lol

File: src/main/java/frc/robot/subsystems/scoring/ScoringSubsystem.java [hunk 2]
```diff
@@ -0,0 +1,732 @@
...
+
+    public void setElevatorPositionSupplier(DoubleSupplier elevatorPositionSupplier) {
+        this.elevatorPositionSupplier = elevatorPositionSupplier;
+    }
+
+    public void setDriveAllignedSupplier(Supplier<Boolean> driveAllignedSupplier) {
+        this.driveAllignedSupplier = driveAllignedSupplier;
+    }
+
+    private boolean willHitStage() {
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/scoring/ScoringSubsystem.java:425]: I think we can remove stage avoidance code for this robot

**Review by jkleiber — COMMENTED** [review]
First pass review. What I've heard from Preston is that we will probably be able to get some intake hardware attached to HITL on Monday to test that the logic works there. We should try to get this PR in a merge-able state + any specific changes to get the conveyor working.

For context, I think this will mostly be a motor + a banner sensor of some kind to demonstrate that the intake automatically runs until a note is detected.

**Comment by jkleiber** [comment]
For clarification on my above message: I think we should merge this PR when it's ready and then make a new PR for the work on integrating hardware on HITL

**Comment by minhnguyenbhs** [comment]
robotcontainer is currently absolutely melted will fix tonight or tomorrow morning

**Comment by minhnguyenbhs** [comment]
@aidnem hi!! sorry I forgot to request your review yesterday night- i think i should be done? possibly?

**Comment by aidnem** [comment]
@minhnguyenbhs no problem! I'm reading over the code but to be honest I'm not confident code reviewing yet so I'll ask Coach Justin tomorrow about it to get the workflow down. This could then be merged at the start of shop tomorrow, as I'll already have read over all the code.

One quick question: Did you resolve his comments on the previous review? I'm new to the review UI but it looks like some of his suggestions haven't been implemented (I could 100% just be reading wrong though).

**Comment by minhnguyenbhs** [comment]
@aidnem I think most of them should be done, although I might have missed something?
oh I did remember specifically not fixing any of the build.gradle stuff because I thought that it would get fixed with the rebase, although that might have been a bad assumption- I'll check on that now haha

File: simgui.json [hunk 3]
```diff
@@ -1,14 +1,23 @@
...
       "/SmartDashboard/Module 3": "Mechanism2d",
+<<<<<<< HEAD
       "/SmartDashboard/PhotonSimField": "Field2d",
       "/SmartDashboard/VisionSystemSim-main/Sim Field": "Field2d"
+=======
+      "/AdvantageKit/RealOutputs/scoring/mechanism2d": "Mechanism2d"
+>>>>>>> f28f4d7 (kind of working i guess??? not fully tested since a bunch of stuff hasn't been included yet:)
+=======
+      "/SmartDashboard/Module 3": "Mechanism2d"
+>>>>>>> 5ca656a (saving for computer restart)
```
Comment (jkleiber) [inline simgui.json:20]: Please resolve merge conflicts in this file

File: src/main/java/frc/robot/BuildConstants.java
Comment (jkleiber) [inline src/main/java/frc/robot/BuildConstants.java:1]: Let's gitignore and untrack this file

File: src/main/java/frc/robot/RobotContainer.java [hunk 4]
```diff
@@ -89,6 +203,18 @@ private void configureSubsystems() {
...
+            intakeSubsystem = new IntakeSubsystem(new IntakeNEOVortex());
+        } else if (Constants.currentMode == Mode.SIM) {
+            scoringSubsystem = new ScoringSubsystem(new ShooterIOSim(), new AimerIOSim());
+            intakeSubsystem = new IntakeSubsystem(new IntakeIOSim());
+        }
+    }
+
+    public void enabledInit() {
+        intakeSubsystem.run(IntakeAction.NONE);
+        scoringSubsystem.setAction(ScoringAction.INTAKE);
```
Comment (jkleiber) [inline src/main/java/frc/robot/RobotContainer.java:217]: On enabled, the scoring subsystem probably shouldn't be attempting any specific action (would expect this to be idle)

File: src/main/java/frc/robot/commands/ShootWithGamepad.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        } else {
+            scoring.setAction(ScoringAction.WAIT);
+        }
+    }
+
+    public void warmupSwitch() {
+        switch (getDriveMode.get()) {
+            case NONE:
+                scoring.setAction(ScoringAction.WAIT);
+                break;
```
Comment (jkleiber) [inline src/main/java/frc/robot/commands/ShootWithGamepad.java:83]: Let's remove the NONE case - if nothing is happening, we should default to the most recently used action

File: src/main/java/frc/robot/constants/Constants.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        RIGHT
+    }
+
+    public static final class FeatureFlags {
+        public static final boolean simulateVision = true;
+        public static final boolean runVision = true;
+        public static final boolean runLocalizer = true;
+
+        public static final boolean runIntake = true;
+        public static final boolean runScoring = true;
```
Comment (jkleiber) [inline src/main/java/frc/robot/constants/Constants.java:33]: These need to be used in RobotContainer still

File: src/main/java/frc/robot/constants/ScoringConstants.java
Comment (jkleiber) [inline src/main/java/frc/robot/constants/ScoringConstants.java:1]: Let's start downselecting the constants in this file to only the ones truly needed by the scoring mechanism. For example, there is no hood on this robot. I'm sure there are other distracting constants as well - let's clean this up

File: src/main/java/frc/robot/subsystems/intake/IntakeIOSim.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    @Override
+    public void setBeltVoltage(double volts) {
+        beltAppliedVolts = MathUtil.clamp(volts, -12.0, 12.0);
+    }
+
+    @Override
+    public void applyOutputs(IntakeIOInputs inputs) {
+        noteInIntakeWheels = true;
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/intake/IntakeIOSim.java:46]: This is sort of an input.

For the sim, there may not be any outputs to apply and that's okay for now

File: src/main/java/frc/robot/subsystems/intake/IntakeNEOVortex.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    @Override
+    public void setIntakeVoltage(double volts) {
+        goalVoltsRightIntake = volts / 12;
+        goalVoltsLeftIntake = volts / 12;
+    }
+
+    @Override
+    public void setBeltVoltage(double volts) {
+        goalVoltsBelt = volts;
+    }
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/intake/IntakeNEOVortex.java:84]: Is this function relevant to this robot?

File: src/main/java/frc/robot/subsystems/intake/IntakeSubsystem.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    @Override
+    public void periodic() {
+        io.updateInputs(inputs);
+        Logger.processInputs("intake", inputs);
+
+        /*Logger.recordOutput("intake/running", inputs.leftIntakeVoltage != 0.0);
+        Logger.recordOutput("intake/belting", inputs.beltVoltage != 0.0);
+
+        Logger.recordOutput("intake/state", state.toString());*/
+
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/intake/IntakeSubsystem.java:36]: We should delete commented out code unless it will be used in the near future

File: src/main/java/frc/robot/subsystems/intake/IntakeSubsystem.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        SmartDashboard.putBoolean("working", true);
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/intake/IntakeSubsystem.java:37]: This debug indication can be deleted

File: src/main/java/frc/robot/subsystems/scoring/AimerIORoboRio.java [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+
+        TalonFXConfigurator aimerRightConfig = aimerRight.getConfigurator();
+        aimerRightConfig.apply(
+                new CurrentLimitsConfigs()
+                        .withStatorCurrentLimit(60)
+                        .withStatorCurrentLimitEnable(true));
+
+        aimerRight.setPosition(0.0);
+
+        controller.setTolerance(0.015);
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/scoring/AimerIORoboRio.java:89]: This tolerance should probably be a constant in ScoringConstants

File: src/main/java/frc/robot/subsystems/scoring/AimerIORoboRio.java [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+
+        aimerLeft.setNeutralMode(NeutralModeValue.Brake);
+        aimerRight.setNeutralMode(NeutralModeValue.Brake);
+
+        aimerRight.setInverted(false);
+
+        TalonFXConfigurator aimerLeftConfig = aimerLeft.getConfigurator();
+        aimerLeftConfig.apply(
+                new CurrentLimitsConfigs()
+                        .withStatorCurrentLimit(60)
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/scoring/AimerIORoboRio.java:78]: Let's make the current limit a constant in ScoringConstants

File: src/main/java/frc/robot/subsystems/scoring/AimerIORoboRio.java [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+
+        State trapezoidSetpoint =
+                profile.calculate(
+                        timer.get(),
+                        new State(initialAngle, initialVelocity),
+                        new State(goalAngleRad, 0));
+
+        double controlSetpoint =
+                MathUtil.clamp(trapezoidSetpoint.position, minAngleClamp, maxAngleClamp);
+        double velocitySetpoint = trapezoidSetpoint.velocity;
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/scoring/AimerIORoboRio.java:205]: Anything involving setpoints or control should be in `applyOutputs`

File: src/main/java/frc/robot/subsystems/scoring/AimerIORoboRio.java [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+            appliedVolts = overrideVolts;
+        } else {
+            double controllerVolts = controller.calculate(getEncoderPosition(), controlSetpoint);
+            appliedVolts =
+                    feedforward.calculate(controlSetpoint, velocitySetpoint) + controllerVolts;
+        }
+
+        appliedVolts = MathUtil.clamp(appliedVolts, -12.0, 12.0);
+
+        Logger.recordOutput("Scoring/motorDisabled", motorDisabled);
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/scoring/AimerIORoboRio.java:221]: Applying voltage is a job for `applyOutputs`

File: src/main/java/frc/robot/subsystems/scoring/AimerIORoboRio.java [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+        inputs.aimSupplyCurrentAmps = aimerRight.getSupplyCurrent().getValueAsDouble();
+    }
+
+    @Override
+    public void applyOutputs(AimerIOInputs inputs) {
+        if (!motorDisabled || override) {
+            aimerRight.setVoltage(inputs.aimAppliedVolts);
+        } else {
+            aimerRight.setVoltage(0.0);
+        }
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/scoring/AimerIORoboRio.java:250]: I think we should move the output computations into this function, and then have this if/else for deciding what volts to apply. If it makes sense, we can have an AimerOutputs class that is autologged and represents the computed control targets.

Maybe we should rename `AimerIOInputs` to `AimerInputs` (and the same for similar other IO interfaces) because it currently is "Aimer Input Output Inputs" which doesn't actually make sense when you think about it. I'd be fine if we did that in a separate PR though since this one is big (please create a ticket for this)

File: src/main/java/frc/robot/subsystems/scoring/AimerIOSim.java [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+        double velocitySetpoint = trapezoidSetpoint.velocity;
+
+        if (override) {
+            appliedVolts = overrideVolts;
+        } else {
+            appliedVolts =
+                    feedforward.calculate(controlSetpoint, velocitySetpoint)
+                            + controller.calculate(sim.getAngleRads(), controlSetpoint);
+            appliedVolts = MathUtil.clamp(appliedVolts, -12.0, 12.0);
+        }
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/scoring/AimerIOSim.java:126]: similar comments for the other IO class - we should have this in `applyOutputs`

Also, if this code is repetitive, we should find a way to put it in a separate file and just use it once. It's better to not maintain two copies of the same code.

File: src/main/java/frc/robot/subsystems/scoring/ShooterIOSim.java [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+                                    shooterLeftSim.getAngularVelocityRadPerSec(),
+                                    shooterLeftGoalVelRPM
+                                            * ConversionConstants.kRPMToRadiansPerSecond);
+            shooterRightAppliedVolts =
+                    shooterFeedforward.calculate(shooterRightSim.getAngularVelocityRadPerSec())
+                            + shooterRightController.calculate(
+                                    shooterRightSim.getAngularVelocityRadPerSec(),
+                                    shooterRightGoalVelRPM
+                                            * ConversionConstants.kRPMToRadiansPerSecond);
+        }
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/scoring/ShooterIOSim.java:91]: Let's move control logic to applyOutputs and migrate any inputs to a new outputs log class

File: src/main/java/frc/robot/subsystems/scoring/ShooterIOTalonFX.java [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+        shooterLeft.getConfigurator().apply(slot0);
+        shooterRight.getConfigurator().apply(slot0);
+    }
+
+    @Override
+    public void updateInputs(ShooterIOInputs inputs) {
+        if (override) {
+            shooterLeft.setVoltage(overrideVolts);
+            shooterRight.setVoltage(overrideVolts);
+        }
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/scoring/ShooterIOTalonFX.java:115]: This should be in applyOutputs since it is setting a voltage

File: src/main/java/frc/robot/subsystems/scoring/ScoringSubsystem.java [hunk 19]
```diff
@@ -0,0 +1,707 @@
...
+
+        aimerInterpolated =
+                new InterpolateDouble(
+                        ScoringConstants.getAimerMap(), 0.0, ScoringConstants.aimMaxAngleRadians);
+
+        timeToPutAimDown = new InterpolateDouble(ScoringConstants.timeToPutAimDownMap(), 0.0, 2.0);
+
+        aimerAvoidElevator =
+                new InterpolateDouble(
+                        ScoringConstants.aimerAvoidElevatorTable(), 0.0, Math.PI / 2.0);
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/scoring/ScoringSubsystem.java:120]: I don't think these are relevant to this robot

**Review by jkleiber — COMMENTED** [review]
Another review pass

**Comment by jkleiber** [comment]
@aidnem how to be a code reviewer is in our docs here: https://team401.github.io/Docs/Programming/Software-Workflow.html#as-the-reviewer

Definitely feel free to add to this if you see me doing something that is worth adding to the list, or if you have additional ideas of how to be a good reviewer. It's possible this deserves its own page (and then we'd just link to that page from this section)

File: src/main/java/frc/robot/RobotContainer.java [hunk 20]
```diff
@@ -89,6 +203,18 @@ private void configureSubsystems() {
...
+            intakeSubsystem = new IntakeSubsystem(new IntakeNEOVortex());
+        } else if (Constants.currentMode == Mode.SIM) {
+            scoringSubsystem = new ScoringSubsystem(new ShooterIOSim(), new AimerIOSim());
+            intakeSubsystem = new IntakeSubsystem(new IntakeIOSim());
+        }
+    }
+
+    public void enabledInit() {
+        intakeSubsystem.run(IntakeAction.NONE);
+        scoringSubsystem.setAction(ScoringAction.WAIT);
```
Comment (jkleiber) [inline src/main/java/frc/robot/RobotContainer.java:217]: these need to be feature flagged

File: src/main/java/frc/robot/RobotContainer.java [hunk 21]
```diff
@@ -89,6 +203,18 @@ private void configureSubsystems() {
             tagVision.setCameraConsumer(
                     (m) -> drive.addVisionMeasurement(m.pose(), m.timestamp(), m.variance()));
         }
+        if (Constants.currentMode == Mode.REAL) {
+            scoringSubsystem = new ScoringSubsystem(new ShooterIOTalonFX(), new AimerIORoboRio());
+            intakeSubsystem = new IntakeSubsystem(new IntakeNEOVortex());
+        } else if (Constants.currentMode == Mode.SIM) {
+            scoringSubsystem = new ScoringSubsystem(new ShooterIOSim(), new AimerIOSim());
+            intakeSubsystem = new IntakeSubsystem(new IntakeIOSim());
```
Comment (jkleiber) [inline src/main/java/frc/robot/RobotContainer.java:211]: These need to be feature flagged

File: src/main/java/frc/robot/subsystems/scoring/ScoringSubsystem.java [hunk 22]
```diff
@@ -0,0 +1,691 @@
...
+        }
+
+        shooterIo.updateInputs(shooterInputs);
+        aimerIo.updateInputs(aimerInputs);
+
+        shooterIo.applyOutputs(shooterOutputs);
+        aimerIo.applyOutputs(aimerOutputs);
+
+        Logger.processInputs("scoring/shooter", shooterInputs);
+        Logger.processInputs("scoring/aimer", aimerInputs);
```
Comment (jkleiber) [inline src/main/java/frc/robot/subsystems/scoring/ScoringSubsystem.java:521]: add processInputs for the outputs

**Review by jkleiber — APPROVED** [review]
approved - please fix the comments in the hardware integration PR
